runtime/parachains/src/inclusion/mod.rs

```diff
@@ -791,6 +950,92 @@ impl<T: Config> Pallet<T> {
+	// FAIL-CI: TODO update docs
+	pub(crate) fn check_upward_messages(
+		config: &HostConfiguration<T::BlockNumber>,
+		para: ParaId,
+		upward_messages: &[UpwardMessage],
+	) -> Result<(), UmpAcceptanceCheckErr> {
+		// Cannot send UMP messages while off-boarding.
+		if <paras::Pallet<T>>::is_offboarding(para) {
+			ensure!(upward_messages.is_empty(), UmpAcceptanceCheckErr::IsOffboarding);
+		}
+
+		let additional_msgs = upward_messages.len();
+		if additional_msgs > config.max_upward_message_num_per_candidate as usize {
+			return Err(UmpAcceptanceCheckErr::MoreMessagesThanPermitted {
+				sent: additional_msgs as u32,
+				permitted: config.max_upward_message_num_per_candidate,
+			})
+		}
+
+		let fp = T::MessageQueue::footprint(AggregateMessageOrigin::Ump(para));
+		let (para_queue_count, mut para_queue_size) = (fp.count, fp.size);
+
+		if para_queue_count
+			.checked_add(additional_msgs as u64)
+			.map(|new| new > config.max_upward_queue_count as u64)
+			.unwrap_or(true)
+		{
+			return Err(UmpAcceptanceCheckErr::CapacityExceeded {
+				count: para_queue_count.saturating_add(additional_msgs as u64),
+				limit: config.max_upward_queue_count as u64,
+			})
+		}
+
+		for (idx, msg) in upward_messages.into_iter().enumerate() {
+			let msg_size = msg.len();
+			if msg_size > config.max_upward_message_size as usize {
+				return Err(UmpAcceptanceCheckErr::MessageSize {
+					idx: idx as u32,
+					msg_size: msg_size as u32,
+					max_size: config.max_upward_message_size,
+				})
+			}
+			// make sure that the queue is not overfilled.
+			// we do it here only once since returning false invalidates the whole relay-chain block.
+			if para_queue_size
+				.checked_add(msg_size as u64)
+				.map(|new| new > config.max_upward_queue_size as u64)
+				.unwrap_or(true)
+			{
+				return Err(UmpAcceptanceCheckErr::TotalSizeExceeded {
+					total_size: para_queue_size.saturating_add(msg_size as u64),
+					limit: config.max_upward_queue_size as u64,
+				})
+			}
+			para_queue_size.saturating_accrue(msg_size as u64);
+		}
+
+		Ok(())
+	}
+
+	/// Enqueues `upward_messages` from a `para`'s accepted candidate block.
+	pub(crate) fn receive_upward_messages(para: ParaId, upward_messages: &[Vec<u8>]) -> Weight {
+		let bounded = upward_messages
+			.iter()
+			.filter_map(|d| BoundedSlice::try_from(&d[..]).ok())
+			.collect();
+		Self::receive_bounded_upward_messages(para, bounded)
+	}
+
+	/// Enqueues storage-bounded `upward_messages` from a `para`'s accepted candidate block.
+	pub(crate) fn receive_bounded_upward_messages(
+		para: ParaId,
+		messages: Vec<BoundedSlice<'_, u8, MaxUmpMessageLen<T>>>,
+	) -> Weight {
+		let count = messages.len() as u32;
+		if count == 0 {
+			return Weight::zero()
+		}
+
+		Self::deposit_event(Event::UpwardMessagesReceived { from: para, count });
+		T::MessageQueue::enqueue_messages(messages.into_iter(), AggregateMessageOrigin::Ump(para));
+		<T as Config>::WeightInfo::receive_upward_messages(count)
```

Comment:
Shouldn't the event technically be fired after this line? I guess it doesn't quite matter as the method isn't fallible...